File: stack/generate_parentheses.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        st = []

        def generate(open: int, close: int):
            if open == close == n:
                res.append("".join(st))
                return

            if open < n:
                st.append("(")
                generate(open + 1, close)
                st.pop()

            if close < open:
                st.append(")")
                generate(open, close + 1)
                st.pop()

        generate(0, 0)
        return res


s = Solution()
print(s.generateParenthesis(3))

File: kotlin/src/main/kotlin/stack/GenerateParentheses.kt
package org.example.stack

fun generateParenthesis(n: Int): List<String> {
    val res = mutableListOf<String>()
    val combination = mutableListOf<Char>()
    fun backtracking(open: Int = 0, close: Int = 0) {
        if (open == n && close == n) {
            res.add(combination.joinToString(""))
            return
        }

        if (open < n) {
            combination.add('(')
            backtracking(open + 1, close)
            combination.removeLast()
        }

        if (close < open) {
            combination.add(')')
            backtracking(open, close + 1)
            combination.removeLast()
        }
    }
    backtracking()
    return res
}

fun main() {
    var input = 3
    var output = generateParenthesis(input)
    println("output: $output, expected: [((())),(()()),(())(),()(()),()()()]")

    input = 1
    output = generateParenthesis(input)
    println("output: $output, expected: [()]")
}